app/controllers/company/CompanyUseSameAddressController.scala

```diff
@@ -65,46 +68,66 @@ class CompanyUseSameAddressController @Inject()(override val messagesApi: Messag
-                case None => request.userAnswers.set(CompanyUseSameAddressPage, false)
-                case Some(resolvedAddress) => request.userAnswers.set(CompanyUseSameAddressPage, value)
-                                      .flatMap(_.set(CompanyAddressPage, resolvedAddress))
-              }
-            } else {
-              request.userAnswers.set(CompanyUseSameAddressPage, value)
+            retrieveTolerantAddress match {
+              case Some(address) =>
+                val updatedUserAnswersTry: Try[UserAnswers] =
+                  if (value) {
+                      getResolvedAddress(address) match {
+                        case None =>
+                          request.userAnswers.set(CompanyUseSameAddressPage, false)
+                        case Some(resolvedAddress) =>
+                          request.userAnswers.set(CompanyUseSameAddressPage, value)
+                            .flatMap(_.set(CompanyAddressPage, resolvedAddress))
+                      }
+                  } else {
+                    request.userAnswers.set(CompanyUseSameAddressPage, value)
+                  }
+                for {
+                  updatedAnswers <- Future.fromTry(updatedUserAnswersTry)
+                  _ <- userAnswersCacheConnector.save(updatedAnswers.data)
+                } yield Redirect(navigator.nextPage(CompanyUseSameAddressPage, NormalMode, updatedAnswers))
+              case _ => Future.successful(Redirect(controllers.routes.SessionExpiredController.onPageLoad()))
             }
-
-            for {
-              updatedAnswers <- Future.fromTry(updatedUserAnswersTry)
-              _ <- userAnswersCacheConnector.save(updatedAnswers.data)
-            } yield Redirect(navigator.nextPage(CompanyUseSameAddressPage, NormalMode, updatedAnswers))
           }
         )
-      }
   }
 
-  private def getJson(form: Form[Boolean])(block: JsObject => Future[Result])(implicit request: DataRequest[AnyContent]): Future[Result] =
-    (BusinessNamePage and ConfirmAddressPage).retrieve.right.map { case companyName ~ address =>
-      val json = Json.obj(
-        "form" -> form,
-        "viewmodel" -> CommonViewModel("company", companyName, routes.CompanyUseSameAddressController.onSubmit().url),
-        "radios" -> Radios.yesNo(form("value")),
-        "address" -> address.lines(countryOptions)
-      )
-      block(json)
+  private def retrieveTolerantAddress(implicit request:DataRequest[_]):Option[TolerantAddress] = {
+    (request.userAnswers.get(AreYouUKCompanyPage),
+      request.userAnswers.get(ConfirmAddressPage),
+      request.userAnswers.get(CompanyRegisteredAddressPage)) match {
+      case (Some(true), Some(address), _) =>
+        Some(address)
+      case (Some(false), _, Some(address)) =>
+        Some(address.toTolerantAddress)
+      case _ => None
+    }
+  }
+
+  private def getJson(form: Form[Boolean])(block: JsObject => Future[Result])(implicit request: DataRequest[AnyContent]): Future[Result] = {
```

Comment:
Same again

Reply:
This is used twice so it's worth having it as a separate method to avoid duplicated.

Reply:
Ok np.